cmd/agent/app/reporter/client_metrics.go

```diff
@@ -0,0 +1,245 @@
+		params.ExpireFrequency = defaultExpireFrequency
+	}
+	if params.ExpireTTL == 0 {
+		params.ExpireTTL = defaultExpireTTL
+	}
+	cm := new(clientMetrics)
+	metrics.MustInit(cm, params.MetricsFactory.Namespace(metrics.NSOptions{Name: "client_stats"}), nil)
+	r := &ClientMetricsReporter{
+		params:        params,
+		clientMetrics: cm,
+		shutdown:      make(chan struct{}),
+		closed:        atomic.NewBool(false),
+	}
+	go r.expireClientMetricsLoop()
+	return r
+}
+
+// EmitZipkinBatch delegates to underlying Reporter.
+func (r *ClientMetricsReporter) EmitZipkinBatch(spans []*zipkincore.Span) error {
+	return r.params.Reporter.EmitZipkinBatch(spans)
+}
+
+// EmitBatch processes client data loss metrics and delegates to the underlying reporter.
+func (r *ClientMetricsReporter) EmitBatch(batch *jaeger.Batch) error {
+	r.updateClientMetrics(batch)
+	return r.params.Reporter.EmitBatch(batch)
+}
+
+// Close stops background gc goroutine for client stats map.
+func (r *ClientMetricsReporter) Close() {
+	if r.closed.CAS(false, true) {
+		close(r.shutdown)
+	}
+}
+
+func (r *ClientMetricsReporter) expireClientMetricsLoop() {
+	ticker := time.NewTicker(r.params.ExpireFrequency)
+	defer ticker.Stop()
+	for {
+		select {
+		case now := <-ticker.C:
+			r.expireClientMetrics(now)
+		case <-r.shutdown:
+			return
+		}
+	}
+}
+
+func (r *ClientMetricsReporter) expireClientMetrics(t time.Time) {
+	var size int64
+	r.lastReceivedClientStats.Range(func(k, v interface{}) bool {
+		stats := v.(*lastReceivedClientStats)
+		stats.lock.Lock()
+		defer stats.lock.Unlock()
+
+		if !stats.lastUpdated.IsZero() && t.Sub(stats.lastUpdated) > r.params.ExpireTTL {
+			r.lastReceivedClientStats.Delete(k)
+			r.params.Logger.Debug("have not heard from a client for a while, freeing stats",
+				zap.Any("client-uuid", k),
+				zap.Time("last-message", stats.lastUpdated),
+			)
+		}
+		size++
+		return true // keep running through all values in the map
+	})
+	r.clientMetrics.ConnectedClients.Update(size)
+}
+
+func (r *ClientMetricsReporter) updateClientMetrics(batch *jaeger.Batch) {
+	clientUUID := clientUUID(batch)
+	if clientUUID == "" {
+		return
+	}
+	if batch.SeqNo == nil {
+		return
+	}
+	entry, found := r.lastReceivedClientStats.Load(clientUUID)
+	if !found {
+		ent, loaded := r.lastReceivedClientStats.LoadOrStore(clientUUID, &lastReceivedClientStats{})
+		if !loaded {
+			r.params.Logger.Debug("received batch from a new client, starting to keep stats",
+				zap.String("client-uuid", clientUUID),
+			)
+		}
+		entry = ent
+	}
+	clientStats := entry.(*lastReceivedClientStats)
+	clientStats.update(*batch.SeqNo, batch.Stats, r.clientMetrics)
+}
+
+func (s *lastReceivedClientStats) update(
+	batchSeqNo int64,
+	stats *jaeger.ClientStats,
+	metrics *clientMetrics,
+) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.batchSeqNo >= batchSeqNo && !wrapped(s.batchSeqNo, batchSeqNo) {
+		// Ignore out of order batches. Once we receive a batch with a larger-than-seen number,
```

Comment:
Would probably be a good idea to add a debug log statement here. While it might not bring much benefit on the usual case, it might be valuable when debugging bad clients (like, a client that always sends the same fixed UUID).

Reply:
Out of order batches are very normal when using UDP, because of unpredictability of the scheduler affecting how agent is processing packets. So I don't think logging it is warranted.

Reply:
They are very normal, hence the debug-level logging. Someone trying to figure out why a specific client metric isn't showing up might benefit from a log message here, in case the client is not behaving appropriately. Or are there any costs in zap for debug-level messages when debug isn't enabled?

Reply:
A general rule for logging in high throughput systems is "don't log things that happen on every request". Out of order batches don't happen on every request, but can easily happen on 50% of requests at high throughput and concurrency. Logging will introduce both runtime overhead (not large, but not zero either), and if debug logging is actually enabled it will flood the logs. At the same time, at low throughput the logging is unlikely to be helpful since out of order batches are also unlikely there.

I'd be ok with adding another metrics counter for out of order batches (but it will be aggregate, not client-specific).

Reply:
Sometimes, I miss HotSpot JVM's cleverness ;-) Not sure a metric would help in this case. Let's leave it out and see if there are requests for it.